core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -740,6 +743,7 @@ private[spark] class TaskSetManager(
     }
     removeRunningTask(tid)
     info.markFinished(state)
+    successfulTaskDurations.remove(taskInfos(tid).duration)
```

Comment:
Ah I see you're saying that the problem is that the check that enough tasks have finished to enable speculation is no longer correct?  If that's the sole issue, why don't we just keep using `tasksSuccessful` for that check rather than looking at the count of items in the heap?